common/events/events_io.cpp

```diff
@@ -31,43 +30,23 @@ AsyncReaderFromReader::~AsyncReaderFromReader() {
 
 error::Error AsyncReaderFromReader::AsyncRead(
 	vector<uint8_t>::iterator start, vector<uint8_t>::iterator end, mio::AsyncIoHandler handler) {
-	if (reader_thread_.joinable()) {
-		return error::Error(
-			make_error_condition(errc::operation_in_progress), "AsyncRead already in progress");
-	}
-
-	auto cancelled = cancelled_;
-	// Expensive to create a thread on every read. Future optimization: Create a thread which
-	// receives work through some channel.
-	reader_thread_ = thread([this, start, end, handler, cancelled]() {
+	// Simple, "cheating" implementation, we just do it synchronously.
+	in_progress_ = true;
+	loop_.Post([this, start, end, handler]() {
 		auto result = reader_->Read(start, end);
-		loop_.Post([this, result, handler, cancelled]() {
-			if (!*cancelled) {
-				// This should always be true, but let's use this as a safeguard
-				// anyway.
-				assert(reader_thread_.joinable());
-				if (reader_thread_.joinable()) {
-					reader_thread_.join();
-				}
-				handler(result);
-			}
-		});
+		in_progress_ = false;
```

Comment:
Wasn't this actually quite the same even before? Spawning a new thread and then doing `loop.Post()` with a blocking `join()` would block the loop until the `read()` is done anyway. Or am I missing something?

Reply:
No, the `Post()` was not done until the `Read()` had finished, so the sequence would be this:

1. `Read()` is called in thread 2, while thread 1 returns and continues processing other events.
2. `Read()` finishes in thread 2, and `Post()` is called.
3. The event loop is running only on thread 1, so it picks up the `Post()`ed task and runs the handler on thread 1.

But now we block instead. And the reason this is ok is, as we talked about, that deep inside the `Read()` we start processing events again, inside the call stack of the tar reader callback.

Reply:
And, just to be clear, that last part with processing inside the callback: I have not added it to this PR yet. I want to test it a bit more first. But it's coming.

Reply:
> No, the `Post()` was not done until the `Read()` had finished

Right! Not so easy to see this in the diff.